src/python/pants/backend/python/lint/pyupgrade/rules.py

```diff
@@ -39,17 +42,39 @@ class PyUpgradeRequest(FixTargetsRequest):
 async def pyupgrade_fix(request: PyUpgradeRequest.Batch, pyupgrade: PyUpgrade) -> FixResult:
     pyupgrade_pex = await Get(VenvPex, PexRequest, pyupgrade.to_pex_request())
 
-    result = await Get(
-        FallibleProcessResult,
-        VenvPexProcess(
-            pyupgrade_pex,
-            argv=(*pyupgrade.args, *request.files),
-            input_digest=request.snapshot.digest,
-            output_files=request.files,
-            description=f"Run pyupgrade on {pluralize(len(request.files), 'file')}.",
-            level=LogLevel.DEBUG,
-        ),
-    )
+    # NB: Pyupgrade isn't idempotent, but eventually converges. So keep running until it stops
+    # changing code. See https://github.com/asottile/pyupgrade/issues/703
+    # (Technically we could not do this. It doesn't break Pants since the next run on the CLI would
+    # use the new file with the new digest. However that isn't the UX we want for our users.)
+    input_digest = request.snapshot.digest
+    for _ in range(10):  # Give the loop an upper bound to guard against inifite runs
```

Comment:
typo in comment... if you have more changes coming..
```suggestion
    for _ in range(10):  # Give the loop an upper bound to guard against infinite runs
```